Add `inner_item_margin_*` Theme constants to the Tree control

joao-pedro-braz opened the pull request:
## Description
This PR adds the "inner_item_margin" Theme constant to the Tree Control.
It behaves like a horizontal padding (in CSS), but only in the active
writing direction (So on LTR it'll apply a left padding and on RTL right
padding).

This PR also makes use of this newly added feature to adjust the Editor Theme.
Previously, icons, like the ones in the FileSystemDock or even in the SceneTree, didn't have proper spacing between borders and would therefore appear too close to them, which was specially noticeable with large Corner Radius values, such as 6.

## Comparisons

### Currently (pre-PR)
![image](https://user-images.githubusercontent.com/37230465/222823334-10a53232-a610-4a4c-b170-bdd9202e2ca9.png)
*File System Dock*

![image](https://user-images.githubusercontent.com/37230465/222824464-4f334f6b-8784-4bd0-bf8f-e8ab890359a3.png)
*File System Dock in RTL*

### With PR
![image](https://user-images.githubusercontent.com/37230465/222823632-ffea51ae-fa8c-43f4-aecb-8699bd864a90.png)
*File System Dock*


![image](https://user-images.githubusercontent.com/37230465/222824309-6ad0d32f-c290-4958-841f-efb7877943ed.png)
*File System Dock in RTL*

Note 2: This superseeds my [first PR](https://github.com/godotengine/godot/pull/74301) and also my [second PR](https://github.com/godotengine/godot/pull/74321)

joao-pedro-braz commented:
Basically https://github.com/godotengine/godot/pull/74321, but on different branches

KoBeWi commented:
The StyleBox does not work well if it draws anything:

https://user-images.githubusercontent.com/2223172/234260819-aa8f9ff9-3fea-4524-b806-89a00fe27515.mp4

It overrides any other style.

If your aim is only to increase spacing, maybe it should be a theme constant instead?

KoBeWi commented:
What I meant is that selected and focused style no longer shows, because item overrides it, which is wrong.
But yeah, constant would be better for this use-case anyway.

KoBeWi commented:
Force-push this one, but you'll need to update commit message and title.

joao-pedro-braz commented:
Sure thing!

joao-pedro-braz committed:
Add the "inner_item_margin" Theme constant to the Tree control

This PR adds the "inner_item_margin" Theme constant to the Tree Control.
It behaves like a horizontal padding (in CSS), but only in the active
writing direction (So on LTR it'll apply a left padding and on RTL right
padding).

The Editor Theme has been updated to make use of this and a result items
in Trees and ItemLists no longer "hugs" their border, expressing a proper
spacing instead.

joao-pedro-braz commented:
@KoBeWi 

Done!

Thank you for the suggestion, this new approach is not only simpler, but also stabler when on RTL, given that the text no longer "moves" around when resizing the item.

joao-pedro-braz commented:
Force-pushed again to update the default value for the editor theme.

joao-pedro-braz commented:
Github apparently had a stroke and force-closed my issue :/
Reopened and fixed a conflict in tree.cpp

KoBeWi reviewed (approved):
The feature itself works correctly and implementation looks ok.
There is no linked proposal though.

joao-pedro-braz commented:
There is now: https://github.com/godotengine/godot-proposals/issues/6902
:smile:

YuriSizov reviewed (changes requested):
> The StyleBox does not work well if it draws anything:

I think we should start accepting cases where only `StyleBoxEmpty` makes sense. We have _pleeeeenty_ of controls that need 4 numeric definitions for each side. Using 4 constants or arbitrary using only one constant doesn't seem practical. Like, why we only add one constant here? And with such vague name too.

If we don't want to go the `StyleBoxEmpty` route, then this PR needs to be updated to introduce 4 constants, like we did in https://github.com/godotengine/godot/pull/70901.

* `item_margin_top`
* `item_margin_bottom`
* `item_margin_left`
* `item_margin_right`

joao-pedro-braz commented:
Interesting, if we were to add the `StyleBoxEmpty` back I guess it could just act as a container for the margin values (ie, without actually being drawn)?
Also, about the name, I got the inspiration from the "item_margin", which does virtually the same thing, but applied to the outside of the item (From the docs: `The horizontal margin at the start of an item. This is used when folding is enabled for the item.`).

joao-pedro-braz commented:
@YuriSizov Went with the 4 constants route, as per https://github.com/godotengine/godot/pull/70901.

joao-pedro-braz commented on scene/gui/tree.cpp:
```diff
@@ -1793,9 +1797,13 @@ void Tree::draw_item_rect(TreeItem::Cell &p_cell, const Rect2i &p_rect, const Co
 	bool rtl = is_layout_rtl();
 
-	int w = 0;
+	rect.set_size(rect.get_size() - Size2i(theme_cache.inner_item_margin_right, theme_cache.inner_item_margin_bottom));
```
Because of RTL that's not necessary.
If we do account for all margin values we get a weird behavior where increasing margin left acts as a margin right increase

kleonc replied:
> Because of RTL that's not necessary.
> If we do account for all margin values we get a weird behavior where increasing margin left acts as a margin right increase

I'd say this indicates that some calculations down below are conceptually incorrect. :upside_down_face:

---

BTW there's `Rect2i::grow_individual` so it could be used instead of manually setting position/size:
```cpp
Rect2i rect = p_rect.grow_individual(-theme_cache.inner_item_margin_left, -theme_cache.inner_item_margin_top, -theme_cache.inner_item_margin_right, -theme_cache.inner_item_margin_bottom);
```

joao-pedro-braz replied:
Thanks @kleonc, updated it to include your suggestion

kleonc commented:
> ### Currently (pre-PR)
> 
> ![image](https://user-images.githubusercontent.com/37230465/222823334-10a53232-a610-4a4c-b170-bdd9202e2ca9.png)
> ### With PR
> 
> ![image](https://user-images.githubusercontent.com/37230465/222823632-ffea51ae-fa8c-43f4-aecb-8699bd864a90.png)

Not sure if it's just me but I do like the "before" look more. The "after" seems like wasting horizontal space for no reason/gain.
What I'm saying is maybe the default editor Theme should be made to look like before (or have a smaller left mergin)? :thinking:
(given nobody mentioned it yet it's likely just me :smile:)

joao-pedro-braz commented:
I think the "wasted space" is an illusion because of that icon in particular (the teardrop rainbow).
With other (specially squared) icons, the issue seems more prominent, not to mention with rounded themes, which was originally my inspiration for this.

YuriSizov commented:
> What I'm saying is maybe the default editor Theme should be made to look like before (or have a smaller left mergin)?

I definitely don't like how it looks now, and it bothered me for years at this point, but I agree that we can use a smaller extra margin by default (and we can perhaps respect the additional spacing editor setting here for a bit of extra space for those who need it).

joao-pedro-braz commented:
Great idea!
I'll work on it rn

joao-pedro-braz commented:
Done!

Without additional spacing:
![image](https://github.com/godotengine/godot/assets/37230465/f21d119d-9d4b-4ff6-a46f-a1bc6bcedf89)

With additional spacing (value of 3):
![image](https://github.com/godotengine/godot/assets/37230465/21e5ccc4-a77f-41c2-a893-d8a300beae77)

YuriSizov commented on editor/editor_themes.cpp:
```diff
@@ -1240,6 +1240,10 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	theme->set_constant("h_separation", "Tree", 6 * EDSCALE);
 	theme->set_constant("guide_width", "Tree", border_width);
 	theme->set_constant("item_margin", "Tree", 3 * default_margin_size * EDSCALE);
+	theme->set_constant("inner_item_margin_bottom", "Tree", 0 * EDSCALE);
+	theme->set_constant("inner_item_margin_left", "Tree", (default_margin_size + extra_spacing) * EDSCALE);
+	theme->set_constant("inner_item_margin_right", "Tree", (default_margin_size + extra_spacing) * EDSCALE);
+	theme->set_constant("inner_item_margin_top", "Tree", 0 * EDSCALE);
```
I think that top and bottom margins can take the extra spacing value into account. Otherwise, `* EDSCALE` needs to be removed from 0s.

YuriSizov commented:
Thanks!
